src/_guides/language/effective-dart/design.md

```diff
@@ -1005,79 +1091,260 @@ var names = people.map((Person person) {
-<?code-excerpt "misc/lib/effective_dart/design_good.dart (avoid-dynamic)"?>
+<?code-excerpt "misc/lib/effective_dart/design_good.dart (prefer-dynamic)"?>
 {% prettify dart %}
-lookUpOrDefault(String name, Map map, defaultValue) {
+T lookUp<T>(String name, Map<String, T> map, T defaultValue) {
   var value = map[name];
   if (value != null) return value;
   return defaultValue;
 }
 {% endprettify %}
 
 {:.bad-style}
-<?code-excerpt "misc/lib/effective_dart/design_bad.dart (avoid-dynamic)"?>
+<?code-excerpt "misc/lib/effective_dart/design_bad.dart (prefer-dynamic)"?>
 {% prettify dart %}
-dynamic lookUpOrDefault(String name, Map map, dynamic defaultValue) {
+lookUp(String name, Map map, defaultValue) {
   var value = map[name];
   if (value != null) return value;
   return defaultValue;
 }
 {% endprettify %}
 
 
-### AVOID annotating with `Function`.
+<aside class="alert alert-info" markdown="1">
+
+Before Dart 2, this guideline stated the exact opposite: *don't* annotate with
+`dynamic` when it is implicit. With the new stronger type system and type
+inference, users now expect Dart to behave like an inferred statically-typed
+language. With that mental model, it is an unpleasant surprise to discover that
+a region of code has silently lost all of the safety and performance of static
+types.
+
+</aside>
+
 
-The `Function` type is barely more precise than using no annotation at all. If
-you're bothering to annotate, it's better to use a precise function type that
-describes the signature and return type of the function.
+### AVOID annotating with only `Function`.
 
-If you are annotating a field, this does mean you have to create a typedef, but
-that's usually worth doing.
+The name `Function` by itself without any return type or parameter signature
+refers to the special `Function` type. This type is only marginally more useful
+than using `dynamic`. If you're going to annotate, prefer a full function type
+that describes the signature and return type of the function.
 
 {:.good-style}
 <?code-excerpt "misc/lib/effective_dart/design_good.dart (avoid-Function)"?>
 {% prettify dart %}
-bool isValidString(String value, bool predicate(String string)) => ...
+bool isValid(String value, bool Function(String string) test) => ...
 {% endprettify %}
 
 {:.bad-style}
 <?code-excerpt "misc/lib/effective_dart/design_bad.dart (avoid-Function)"?>
 {% prettify dart %}
-bool isValidString(String value, Function predicate) => ...
+bool isValid(String value, Function test) => ...
+{% endprettify %}
+
+In Dart 1, users sometimes used `Function` to avoid the necessary tedium of
+creating a typedef to use the function type in a field, variable, or generic
+type argument. Dart 2 has a [general function type syntax][fn syntax] that can
+be used anywhere a type annotation can occur, so this shortcut is no longer
+needed.
+
+[fn syntax]: #prefer-inline-function-types-over-typedefs
+
+One exception to this guideline is if you want a type that represents the union
```

Comment:
Since we now have a "PREFER annotating with `dynamic` ..." guideline, I think that it would make sense to pull out this "one exception" into its own guideline.

> ### PREFER annotation with `Function` only ...

Maybe complete with something along the lines of "... when the function arity isn't fixed/know"?

Reply:
That feels like overkill to me. This guideline is already a "PREFER" one, which implies there are exceptions and other PREFER/AVOID guidelines sometimes spell out those exceptions too.

Reply:
Ok